src/part-guide/adv-async-await.md

```diff
@@ -2,26 +2,142 @@
+
+Similarly to async functions, an async block is a deferred version of a regular block. An async block scopes code and names together, but at runtime it is not immediately executed and evaluates to a future. To execute the block and obtain the result, it must be `await`ed. E.g.,
+
+```rust,norun
+let s1 = {
+    let a = 42;
+    format!("The answer is {a}")
+};
+
+let s2 = async {
+    let q = question().await;
+    format!("The question is {q}")
+};
+```
+
+If we were to execute this snippet, `s1` would be a string which could be printed, but `s2` would be a future - `question()` would not have been called. To print `s2`, we first have to `s2.await`.
+
+An async block is the simplest way to create a future, and the simplest way to create an async context for deferred work.
+
+Unfortunately, control flow with async blocks is a little quirky. Because an async block creates a future rather than straightforwardly executing, it behaves more like a function than a regular block with respect to control flow. `break` and `continue` cannot go 'through' an async block like they can with regular blocks, instead you have to use `return`:
+
+```rust,norun
+loop {
+    {
+        if ... {
+            // ok
+            continue;
+        }
+    }
+
+    async {
+        if ... {
+            // not ok
+            // continue;
+
+            // ok - continues with the next execution of the `loop`
+            return;
+        }
+    }.await
+}
+```
+
+To implement `break` you would need to test the value of the block.
+
+Likewise, `?` inside an async block will terminate execution of the future in the presence of an error, causing the `await`ed block to take the value of the error, but won't exit the surrounding function (like `?` in a regular block would). You'll need another `?` after `await` for that:
+
+```rust,norun
+async {
+    let x = foo()?;   // This `?` only exits the async block, not the surrounding function.
+    consume(x);
+    Ok(())
+}.await?
+```
+
+Annoyingly, this often confuses the compiler since (unlike functions) the 'return' type of an async block is not explicitly stated. You'll probably need to add some type annotations on variables or use turbofish types to make this work, e.g., `Ok::<_, MyError>(())` instead of `Ok(())` in the above example.
+
+A function which returns an async block is pretty similar to an async function. Writing `async fn foo() -> ... { ... }` is roughly equivalent to `fn foo() -> ... { async { ... } }`. In fact, from the caller's perspective they are equivalent, and changing from one form to the other is not a breaking change. Furthermore, you can override one with the other when implementing an async trait (see below). However, you do have to adjust the type, making the `Future` explicit in the async block version: `async fn foo() -> Foo` becomes `fn foo() -> impl Future<Output = Foo>` (you might also need to make other bounds explicit, e.g., `Send` and `'static`).
```

Comment:
In Rust 2024, due to the changes to lifetime capture rules, these should be exactly equivalent.  In either case, auto traits like `Send` are going to leak to the returned opaque future.

It'd probably be more accurate to say that there are times when you want or need to desugar to the `-> impl Future<..>` form so that you can specify bounds on that returned opaque type rather than saying that one needs to adjust those things when converting between the forms.

Reply:
Tracked for later...